File: apps/web/src/lib/api/retention.ts
import { apiClient } from "@/lib/api/client";

export interface PitchComment {
  id: string;
  startup_profile_id: string;
  user_id: string;
  user_name: string;
  user_role: string;
  parent_id: string | null;
  content: string;
  created_at: string;
}

export interface PipelineItem {
  id: string;
  user_id: string;
  target_user_id: string;
  target_name: string;
  target_email: string;
  target_role: string;
  target_subtitle?: string | null;
  stage: "matched" | "intro_sent" | "meeting_scheduled" | "due_diligence" | "term_sheet" | "closed" | "passed";
  notes?: string | null;
  created_at: string;
  updated_at: string;
}

export interface InvestorUpdate {
  id: string;
  startup_profile_id: string;
  startup_name: string;
  title: string;
  month_year: string;
  mrr?: number | null;
  runway_months?: number | null;
  highlights: string;
  lowlights?: string | null;
  asks?: string | null;
  is_public: boolean;
  created_at: string;
}

export interface WatchlistItem {
  id: string;
  user_id: string;
  target_type: "startup" | "investor";
  target_id: string;
  title: string;
  subtitle?: string | null;
  created_at: string;
}

export interface NotificationItem {
  id: string;
  user_id: string;
  type: string;
  title: string;
  message: string;
  link_url?: string | null;
  is_read: boolean;
  created_at: string;
}

export interface EcosystemInsights {
  total_startups: number;
  total_investors: number;
  avg_valuation_inr: number;
  avg_funding_required_inr: number;
  top_industries: { industry: string; count: number }[];
  active_matches_count: number;
}

// ── Comments API ─────────────────────────────────────────────────────
export async function getPitchComments(startupProfileId: string) {
  const res = await apiClient.get<{ items: PitchComment[]; total: number }>(
    `/pitches/${startupProfileId}/comments`
  );
  return res.data;
}

export async function addPitchComment(startupProfileId: string, content: string, parentId?: string) {
  const res = await apiClient.post<PitchComment>(
    `/pitches/${startupProfileId}/comments`,
    { content, parent_id: parentId }
  );
  return res.data;
}

// ── Pipeline API ─────────────────────────────────────────────────────
export async function getPipeline() {
  const res = await apiClient.get<{ items: PipelineItem[] }>("/pipeline");
  return res.data;
}

export async function addPipelineItem(targetUserId: string, stage = "matched", notes?: string) {
  const res = await apiClient.post<PipelineItem>("/pipeline", {
    target_user_id: targetUserId,
    stage,
    notes,
  });
  return res.data;
}

export async function updatePipelineStage(itemId: string, stage: string, notes?: string) {
  const res = await apiClient.patch<PipelineItem>(`/pipeline/${itemId}`, {
    stage,
    notes,
  });
  return res.data;
}

export async function deletePipelineItem(itemId: string) {
  await apiClient.delete(`/pipeline/${itemId}`);
}

// ── Investor Updates API ─────────────────────────────────────────────
export async function getInvestorUpdates(startupProfileId: string) {
  const res = await apiClient.get<{ items: InvestorUpdate[] }>(
    `/startup-profile/${startupProfileId}/updates`
  );
  return res.data;
}

export async function createInvestorUpdate(payload: {
  title: string;
  month_year: string;
  mrr?: number;
  runway_months?: number;
  highlights: string;
  lowlights?: string;
  asks?: string;
}) {
  const res = await apiClient.post<InvestorUpdate>("/startup-profile/me/updates", payload);
  return res.data;
}

// ── Watchlist API ────────────────────────────────────────────────────
export async function getWatchlist() {
  const res = await apiClient.get<{ items: WatchlistItem[] }>("/watchlist");
  return res.data;
}

export async function toggleWatchlist(targetType: "startup" | "investor", targetId: string) {
  const res = await apiClient.post<{ saved: boolean; message: string }>("/watchlist/toggle", {
    target_type: targetType,
    target_id: targetId,
  });
  return res.data;
}

// ── Notifications API ────────────────────────────────────────────────
export async function getNotifications() {
  const res = await apiClient.get<{ items: NotificationItem[]; unread_count: number }>(
    "/notifications"
  );
  return res.data;
}

export async function markAllNotificationsRead() {
  const res = await apiClient.post<{ message: string }>("/notifications/read-all");
  return res.data;
}

// ── Ecosystem Insights API ───────────────────────────────────────────
export async function getEcosystemInsights() {
  const res = await apiClient.get<EcosystemInsights>("/insights");
  return res.data;
}

File: apps/api/app/schemas/retention.py
from datetime import datetime
from uuid import UUID
from pydantic import BaseModel, Field


# ── Comments ──────────────────────────────────────────────────────────
class PitchCommentCreate(BaseModel):
    content: str = Field(..., min_length=1, max_length=2000)
    parent_id: UUID | None = None


class PitchCommentPublic(BaseModel):
    id: UUID
    startup_profile_id: UUID
    user_id: UUID
    user_name: str
    user_role: str
    parent_id: UUID | None = None
    content: str
    created_at: datetime

    class Config:
        from_attributes = True


class PitchCommentListResponse(BaseModel):
    items: list[PitchCommentPublic]
    total: int


# ── Pipeline ──────────────────────────────────────────────────────────
class PipelineItemCreate(BaseModel):
    target_user_id: UUID
    stage: str = "matched"  # matched, intro_sent, meeting_scheduled, due_diligence, term_sheet, closed, passed
    notes: str | None = None


class PipelineItemUpdate(BaseModel):
    stage: str
    notes: str | None = None


class PipelineItemPublic(BaseModel):
    id: UUID
    user_id: UUID
    target_user_id: UUID
    target_name: str
    target_email: str
    target_role: str
    target_subtitle: str | None = None  # Industry or Org name
    stage: str
    notes: str | None = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PipelineListResponse(BaseModel):
    items: list[PipelineItemPublic]


# ── Investor Updates ──────────────────────────────────────────────────
class InvestorUpdateCreate(BaseModel):
    title: str = Field(..., min_length=3, max_length=255)
    month_year: str = Field(..., min_length=3, max_length=50)
    mrr: float | None = None
    runway_months: int | None = None
    highlights: str = Field(..., min_length=5)
    lowlights: str | None = None
    asks: str | None = None
    is_public: bool = True


class InvestorUpdatePublic(BaseModel):
    id: UUID
    startup_profile_id: UUID
    startup_name: str
    title: str
    month_year: str
    mrr: float | None = None
    runway_months: int | None = None
    highlights: str
    lowlights: str | None = None
    asks: str | None = None
    is_public: bool
    created_at: datetime

    class Config:
        from_attributes = True


class InvestorUpdateListResponse(BaseModel):
    items: list[InvestorUpdatePublic]


# ── Watchlist ─────────────────────────────────────────────────────────
class WatchlistItemCreate(BaseModel):
    target_type: str  # "startup" or "investor"
    target_id: UUID


class WatchlistItemPublic(BaseModel):
    id: UUID
    user_id: UUID
    target_type: str
    target_id: UUID
    title: str
    subtitle: str | None = None
    created_at: datetime


class WatchlistListResponse(BaseModel):
    items: list[WatchlistItemPublic]


# ── Notifications ─────────────────────────────────────────────────────
class NotificationPublic(BaseModel):
    id: UUID
    user_id: UUID
    type: str
    title: str
    message: str
    link_url: str | None = None
    is_read: bool
    created_at: datetime


class NotificationListResponse(BaseModel):
    items: list[NotificationPublic]
    unread_count: int


# ── Insights ──────────────────────────────────────────────────────────
class EcosystemInsightsResponse(BaseModel):
    total_startups: int
    total_investors: int
    avg_valuation_inr: float
    avg_funding_required_inr: float
    top_industries: list[dict[str, int | str]]
    active_matches_count: int


# ── Fantasy Angel Portfolio Game ──────────────────────────────────────
class VirtualInvestmentCreate(BaseModel):
    startup_profile_id: UUID
    amount: float = Field(default=100000.0, gt=0)


class VirtualInvestmentPublic(BaseModel):
    id: UUID
    startup_profile_id: UUID
    startup_name: str
    industry: str
    amount: float
    current_value: float
    return_percent: float
    created_at: datetime


class FantasyPortfolioResponse(BaseModel):
    available_credits: float
    total_invested: float
    current_portfolio_value: float
    net_return_percent: float
    investments: list[VirtualInvestmentPublic]
